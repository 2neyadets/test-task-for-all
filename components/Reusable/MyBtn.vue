<template lang="pug">
  button.my-btn(
    @click="$emit('click')"
    :ref="'button' + id"
    :type="type"
    role="button"
    tabindex="0"
    :class="btnClass"
    :disabled="disable"
  )
    .full-height.row.items-center(:style="`font-size: ${size}px;`")
      template(v-if="icon")
        .col-auto.flex.flex-center
          i.material-icons(
            :class="label ? 'margin-right' : ''"
          ) {{icon}}
      .col {{label}}
</template>

<script>
import { getId } from '../../assets/helpers'

export default {
  name: 'MyBtn',
  components: {},
  props: {
    type: {
      type: String,
      default: 'button',
    },
    label: {
      type: String,
      default: '',
    },
    icon: {
      type: String,
      default: '',
    },
    size: {
      type: [String, Number],
      default: 20,
    },
    round: {
      type: Boolean,
      default: false,
    },
    active: {
      type: Boolean,
      default: false,
    },
    flat: {
      type: Boolean,
      default: false,
    },
    disable: {
      type: Boolean,
      default: false,
    },
    autofocus: {
      type: Boolean,
      default: false,
    },
  },
  data () {
    return {
      id: getId(),
    }
  },
  computed: {
    btnClass () {
      return {
        'my-btn--hoverable': !this.disable,
        'my-btn--flat': this.flat,
        'my-btn--round': this.round,
        'my-btn--active': this.active,
      }
    },
  },
  watch: {},
  created () {},
  mounted () {
    if (this.autofocus) this.$refs['button' + this.id].focus()
  },
  updated () {},
  methods: {},
}
</script>

<style scoped lang="stylus">
  .my-btn
    display inline-flex
    flex-direction column
    align-items stretch
    position relative
    outline 0
    border 1px solid white
    vertical-align middle
    padding 6px
    font-size 14px
    line-height 1.715em
    text-decoration none
    color white
    background transparent
    font-weight 500
    text-transform uppercase
    text-align center
    width auto
    height auto
    transition .5s

    &--hoverable:hover, &--hoverable:focus
      color black
      background white
    &--flat
      border 0
      &:hover, &:focus
        color inherit
        background rgba(255, 255, 255, .15)
    &--round
      border-radius 50%
    &--active
      color black
      background white
  .margin-right
    margin-right 12px
</style>
