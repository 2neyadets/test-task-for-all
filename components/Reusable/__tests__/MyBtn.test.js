import Vuex from 'vuex'
import { shallowMount, createLocalVue } from '@vue/test-utils'
import MyBtn from '../MyBtn'

const localVue = createLocalVue()
localVue.use(Vuex)

describe('MyBtn component', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(MyBtn)
  })

  it('mounts without errors', done => {
    localVue.config.errorHandler = done // для отлавливания ошибок внутри nextTick
    expect.assertions(1)
    expect(wrapper).toBeTruthy()
    done()
  })

  it('can be round, active, flat, hoverable', done => {
    wrapper.setProps({
      round: true,
      active: true,
      flat: true,
      disable: true,
    })
    localVue.nextTick(() => {
      expect.assertions(4)
      expect(wrapper.classes('my-btn--round')).toBe(true)
      expect(wrapper.classes('my-btn--active')).toBe(true)
      expect(wrapper.classes('my-btn--flat')).toBe(true)
      expect(wrapper.classes('my-btn--hoverable')).toBe(false)
      done()
    })
  })

  it('can be disabled', done => {
    wrapper.setProps({
      disable: true,
    })
    localVue.nextTick(() => {
      expect.assertions(1)
      expect(wrapper.attributes('disabled')).toBe('disabled')
      done()
    })
  })

  it('can change type', done => {
    wrapper.setProps({
      type: 'submit',
    })
    localVue.nextTick(() => {
      expect.assertions(1)
      expect(wrapper.attributes('type')).toBe('submit')
      done()
    })
  })

  it('can autofocus', () => {
    const localWrapper = shallowMount(MyBtn, {
      propsData: {
        autofocus: true,
      },
    })
    expect.assertions(1)
    expect(localWrapper.element).toBe(document.activeElement)
  })

  it('can change size', done => {
    expect.assertions(2)
    wrapper.setProps({
      size: 35,
    })
    localVue.nextTick(() => {
      expect(wrapper.find('.full-height').attributes().style).toBe('font-size: 35px;')
      done()
    })
    wrapper.setProps({
      size: '35',
    })
    localVue.nextTick(() => {
      expect(wrapper.find('.full-height').attributes().style).toBe('font-size: 35px;')
      done()
    })
  })

  it('can set label', done => {
    const label = 'Привет!'
    wrapper.setProps({
      label
    })
    localVue.nextTick(() => {
      expect.assertions(1)
      expect(wrapper.html()).toContain(label)
      done()
    })
  })

  it('can set icon', done => {
    const icon = 'cancel'
    wrapper.setProps({
      icon,
    })
    localVue.nextTick(() => {
      expect.assertions(2)
      expect(wrapper.html()).toContain(icon)
      expect(wrapper.find('i').classes('margin-right')).toBe(false)
      done()
    })
  })

  it('can set label and icon together', done => {
    const icon = 'add',
      label = 'Добавить'
    wrapper.setProps({
      icon,
      label,
    })
    localVue.nextTick(() => {
      expect.assertions(3)
      expect(wrapper.find('i').element.tagName).toBe('I')
      expect(wrapper.find('i').classes('margin-right')).toBe(true)
      expect(wrapper.html()).toContain(icon)
      done()
    })
  })

  it('click works', () => {
    wrapper.trigger('click')
    expect.assertions(1)
    expect(wrapper.emitted('click').length).toBe(1)
  })

  it('match snapshot', () => {
    expect.assertions(1)
    expect(wrapper.element).toMatchSnapshot()
  })
})
