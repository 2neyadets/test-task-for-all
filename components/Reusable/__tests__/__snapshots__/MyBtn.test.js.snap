// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyBtn component match snapshot 1`] = `
<button
  class="my-btn my-btn--hoverable"
  role="button"
  tabindex="0"
  type="button"
>
  <div
    class="full-height row items-center"
    style="font-size: 20px;"
  >
    <!---->
    <div
      class="col"
    >
      
    </div>
  </div>
</button>
`;
